tests/test_run_reference_vmm.py

```diff
@@ -93,33 +95,59 @@ def shutdown(vmm_process):
     # TimeoutExpired exception and fail the test.
     vmm_process.wait(timeout=3)
 
+def setup_stdout_nonblocking(vmm_process):
+
+    ## We'll need to do non-blocking I/O with the underlying sub-process since
+    # we cannot use `communicate`, because `communicate` would close the
+    # `stdin` that we later want to use to `shutdown`, to do that by hand,
+    # we set `vmm_process.stdout` to non-blocking
+    # Then we can use `os.read` that would raise `BlockingIOError`
+
+    # FIXME: This should NOT be required to be done on every call, do it when we
+    #        'Class'ify the test case
+    flags = fcntl.fcntl(vmm_process.stdout, fcntl.F_GETFL)
+    fcntl.fcntl(vmm_process.stdout, fcntl.F_SETFL, flags | os.O_NONBLOCK)
+
+def expect_string(vmm_process, expected_string, timeout=TEST_TIMEOUT):
+
+    setup_stdout_nonblocking(vmm_process)
+
+    # No. of seconds after which we'll give up
+    giveup_after = timeout
+    then = time.time()
+    found = False
+    # This is required because the pattern we are expecting might get split across two reads
+    all_data = bytes()
+    while not found:
+        try:
+            x = os.read(vmm_process.stdout.fileno(), 4096)
+            all_data += x
+            for line in all_data.split(b'\r\n'):
+                if expected_string in line.decode():
+                    return line.decode()
+                    found = True
+            # Whatever remains is collected in `all_data`.
+            all_data = line
+        except Exception as e:
+            time.sleep(1)
```

Comment:
Why do we sleep here?

Reply:
The exception is [BlockingIOError](https://docs.python.org/3/library/exceptions.html#BlockingIOError). This is like `EWOULDBLOCK` so retrying immediately would likely throw the exception again, hence the idea is to sleep for a second and try. I could have done this with `select` and timeout as well.